File: ios/Classes/AudioPlugin.m
#import "AudioPlugin.h"
#import "AudioPlayer.h"

#import <UIKit/UIKit.h>
#import <AVKit/AVKit.h>
#import <AudioToolbox/AudioToolbox.h>
#import <AVFoundation/AVFoundation.h>

static FlutterMethodChannel* channel;
static NSMutableDictionary* players;

@implementation AudioPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar
{
    channel = [FlutterMethodChannel methodChannelWithName:@"audio" binaryMessenger:[registrar messenger]];
    AudioPlugin* instance = [[AudioPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result
{
    NSString* uid = call.arguments[@"uid"];

    NSLog(@"handleMethodCall: [%@] %@", call.method, uid);

    for(NSString *key in [call.arguments allKeys]) {
        NSLog(@"handleMethodCall ARG=%@, VALUE=%@", key, [call.arguments objectForKey:key]);
    }

    if (![players objectForKey:uid])
        players[uid] = [[AudioPlayer alloc] initWithUid:uid channel:channel];

    if ([@"player.play" isEqualToString:call.method])
    {
        [players[uid] play:call.arguments[@"url"]];
        result(nil);
    }
    else if ([@"player.preload" isEqualToString:call.method])
    {
        [players[uid] preload:call.arguments[@"url"]];
        result(nil);
    }
    else if ([@"player.pause" isEqualToString:call.method])
    {
        [players[uid] pause];
        result(nil);
    }
    else if ([@"player.stop" isEqualToString:call.method])
    {
        [players[uid] stop:call.arguments[@"completed"]];
        result(nil);
    }
    else if ([@"player.seek" isEqualToString:call.method])
    {
        NSLog(@"position = %.20f", [call.arguments[@"position"] doubleValue]);
        [players[uid] seek:[call.arguments[@"position"] doubleValue]];
        result(nil);
    }
    else if ([@"player.release" isEqualToString:call.method])
    {
        [players[uid] releasePlayer];
        result(nil);
    }
    else
    {
        result(FlutterMethodNotImplemented);
    }
}

- (id)init
{
    self = [super init];
    
    if (self)
    {
        players = [[NSMutableDictionary alloc] init];
    }

    return self;
}

@end

File: android/src/main/java/com/idofilus/audio/AudioPlugin.java
package com.idofilus.audio;

import java.util.WeakHashMap;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * AudioPlugin
 */
public class AudioPlugin implements MethodCallHandler
{
    private static final String TAG = AudioPlugin.class.getName();

    private MethodChannel channel;
    private static WeakHashMap<String, AudioPlayer> players = new WeakHashMap<>();

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar)
    {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "audio");
        channel.setMethodCallHandler(new AudioPlugin(registrar, channel));
    }

    public AudioPlugin(Registrar registrar, MethodChannel channel)
    {
        this.channel = channel;
    }

    @Override
    public void onMethodCall(MethodCall call, Result result)
    {
        String uid = call.argument("uid");

        System.out.println(String.format("onMethodCall method=%s uid=%s", call.method, uid));

        // Make sure we have the audio player ready
        initialize(uid);

        switch (call.method)
        {
            case "player.play":
                players.get(uid).play((String)call.argument("url"), (int)call.argument("positionInterval"));
                result.success(null);
                break;

            case "player.preload":
                players.get(uid).preload((String)call.argument("url"), (int)call.argument("positionInterval"));
                result.success(null);
                break;

            case "player.pause":
                players.get(uid).pause();
                result.success(null);
                break;

            case "player.stop":
                players.get(uid).stop(false);
                result.success(null);
                break;

            case "player.seek":
                players.get(uid).seek((double) call.argument("position"));
                result.success(null);
                break;

            case "player.release":
                players.get(uid).release();
                result.success(null);
                break;

            default:
                result.notImplemented();
                break;
        }
    }

    /// Initialize the media player
    private void initialize(final String uid)
    {
        if (players.containsKey(uid))
            return;

        players.put(uid, new AudioPlayer(channel, uid));
    }
}
